File: src/adapters/csv-import.ts
import { readFile } from "node:fs/promises";
import { CSVRowError, CSVSchemaError, EmptyCSVError } from "../errors.js";
import { sumCost } from "../schema.js";
import type {
  AdapterResult,
  DateWindow,
  ToolId,
  UserUsage,
} from "../schema.js";

const EXPECTED_COLUMNS = ["date", "user_email", "cost_usd", "is_estimated"];

interface CSVRow {
  date: string;
  user_email: string;
  cost_usd: string;
  is_estimated: string;
}

/**
 * Strips C0 control characters (0x00-0x1f), including ANSI/OSC terminal
 * escape sequences, from a CSV cell value. Without this, a crafted
 * user_email in an imported CSV could inject escape codes into the
 * non-JSON terminal summary output (found during the [redacted ]review) -
 * JSON.stringify already escapes these, so this only matters for the
 * human-readable render path, but stripping at parse time protects every
 * consumer, not just that one.
 */
function stripControlChars(value: string): string {
  return value.replace(/[\x00-\x1f]/g, "");
}

function parseCSV(text: string): CSVRow[] {
  const lines = text
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0);

  if (lines.length === 0) {
    throw new EmptyCSVError("<csv content>");
  }

  const header = lines[0]?.split(",").map((h) => h.trim()) ?? [];
  const missingColumns = EXPECTED_COLUMNS.filter(
    (col) => !header.includes(col),
  );
  if (missingColumns.length > 0) {
    throw new CSVSchemaError(EXPECTED_COLUMNS);
  }

  const dateIdx = header.indexOf("date");
  const emailIdx = header.indexOf("user_email");
  const costIdx = header.indexOf("cost_usd");
  const estimatedIdx = header.indexOf("is_estimated");

  return lines.slice(1).map((line) => {
    const cells = line.split(",").map((c) => stripControlChars(c.trim()));
    return {
      date: cells[dateIdx] ?? "",
      user_email: cells[emailIdx] ?? "",
      cost_usd: cells[costIdx] ?? "",
      is_estimated: cells[estimatedIdx] ?? "",
    };
  });
}

/**
 * Imports before-window spend from a CSV file for a tool whose admin API
 * doesn't cover the requested window (e.g. Claude Code before 2026-01-01).
 * Schema: date, user_email, cost_usd, is_estimated. One row per user per day.
 */
export async function importFromCSV(
  csvPath: string,
  source: ToolId,
  window: DateWindow,
): Promise<AdapterResult> {
  const text = await readFile(csvPath, "utf-8");
  if (text.trim().length === 0) {
    throw new EmptyCSVError(csvPath);
  }

  const rows = parseCSV(text);
  const userTotals = new Map<string, UserUsage>();

  rows.forEach((row, index) => {
    const rowNumber = index + 2; // +1 for header row, +1 for 1-based numbering
    if (row.user_email.length === 0) {
      throw new CSVRowError(rowNumber, "user_email is empty");
    }

    const cost = Number.parseFloat(row.cost_usd);
    if (!Number.isFinite(cost)) {
      throw new CSVRowError(rowNumber, `cost_usd "${row.cost_usd}" is not a valid number`);
    }

    const existing = userTotals.get(row.user_email);
    const isEstimated = row.is_estimated.toLowerCase() === "true";

    if (existing) {
      existing.costUsd += cost;
      existing.isEstimated = existing.isEstimated || isEstimated;
    } else {
      userTotals.set(row.user_email, {
        userId: row.user_email,
        userEmail: row.user_email,
        inputTokens: null,
        outputTokens: null,
        cacheReadTokens: null,
        cacheWriteTokens: null,
        requests: null,
        costUsd: cost,
        isEstimated,
      });
    }
  });

  const users = [...userTotals.values()];
  return {
    source,
    window,
    totalCostUsd: sumCost(users),
    isEstimated: users.some((u) => u.isEstimated),
    users,
  };
}

File: python/src/teamspend/adapters/csv_import.py
"""
CSV-import fallback adapter, for the history a live admin API can't reach.

Ported from src/adapters/csv-import.ts. Schema: date, user_email, cost_usd,
is_estimated. One row per user per day; rows are aggregated per user_email.

Uses a simple split(",") parser rather than the stdlib `csv` module, on
purpose: the TypeScript original does the same (no quoted-field or embedded-
comma support), and keeping the two implementations' exact parsing
semantics identical matters more here than gaining stdlib `csv`'s richer
quoting support that the npm CLI doesn't have either.
"""
from __future__ import annotations

import re
from pathlib import Path
from typing import Dict, List

from ..errors import CSVRowError, CSVSchemaError, EmptyCSVError
from ..types import AdapterResult, DateWindow, ToolId, UserUsage, sum_cost

EXPECTED_COLUMNS = ["date", "user_email", "cost_usd", "is_estimated"]

_CONTROL_CHAR_RE = re.compile(r"[\x00-\x1f]")


def _strip_control_chars(value: str) -> str:
    """
    Strips C0 control characters (0x00-0x1f), including ANSI/OSC terminal
    escape sequences, from a CSV cell value. Without this, a crafted
    user_email in an imported CSV could inject escape codes into the
    non-JSON terminal summary output.
    """
    return _CONTROL_CHAR_RE.sub("", value)


def _parse_csv(text: str) -> List[Dict[str, str]]:
    lines = [line.strip() for line in text.split("\n") if line.strip()]
    if not lines:
        raise EmptyCSVError("<csv content>")

    header = [h.strip() for h in lines[0].split(",")]
    missing_columns = [col for col in EXPECTED_COLUMNS if col not in header]
    if missing_columns:
        raise CSVSchemaError(EXPECTED_COLUMNS)

    date_idx = header.index("date")
    email_idx = header.index("user_email")
    cost_idx = header.index("cost_usd")
    estimated_idx = header.index("is_estimated")

    rows: List[Dict[str, str]] = []
    for line in lines[1:]:
        cells = [_strip_control_chars(c.strip()) for c in line.split(",")]

        def cell(idx: int) -> str:
            return cells[idx] if idx < len(cells) else ""

        rows.append(
            {
                "date": cell(date_idx),
                "user_email": cell(email_idx),
                "cost_usd": cell(cost_idx),
                "is_estimated": cell(estimated_idx),
            }
        )
    return rows


def import_from_csv(csv_path: str, source: ToolId, window: DateWindow) -> AdapterResult:
    """
    Imports before-window spend from a CSV file for a tool whose admin API
    doesn't cover the requested window (e.g. Claude Code before 2026-01-01).
    """
    text = Path(csv_path).read_text(encoding="utf-8")
    if not text.strip():
        raise EmptyCSVError(csv_path)

    rows = _parse_csv(text)
    user_totals: Dict[str, UserUsage] = {}

    for index, row in enumerate(rows):
        row_number = index + 2  # +1 for header row, +1 for 1-based numbering
        if len(row["user_email"]) == 0:
            raise CSVRowError(row_number, "user_email is empty")

        try:
            cost = float(row["cost_usd"])
        except ValueError:
            raise CSVRowError(
                row_number, f'cost_usd "{row["cost_usd"]}" is not a valid number'
            )

        existing = user_totals.get(row["user_email"])
        is_estimated = row["is_estimated"].lower() == "true"

        if existing:
            existing.cost_usd += cost
            existing.is_estimated = existing.is_estimated or is_estimated
        else:
            user_totals[row["user_email"]] = UserUsage(
                user_id=row["user_email"],
                user_email=row["user_email"],
                input_tokens=None,
                output_tokens=None,
                cache_read_tokens=None,
                cache_write_tokens=None,
                requests=None,
                cost_usd=cost,
                is_estimated=is_estimated,
            )

    users_list = list(user_totals.values())
    return AdapterResult(
        source=source,
        window=window,
        total_cost_usd=sum_cost(users_list),
        is_estimated=any(u.is_estimated for u in users_list),
        users=users_list,
    )
